readme.md

```diff
@@ -0,0 +1,119 @@
+
+This is realy like minimal version of the app. possible next steps what should be done are described in _To do (/nice to have)_ section
+### Description
+
+- App should take pipeline flow from .yaml file and inputs from command line.
+- Check, if the declaration is correct, validate if it can finish.
+- Execute and return results
+
+### Guide
+
+Before you start install requirements
+
+```
+pip install -r requirements.txt
+```
+
+Run pipeline principle demonstrator:
+```
+python async_pipeline_principle.py
+```
+
+Run client with assignment defined yaml:
+```
+python pipeline_cli.py --file pipeline_assignment.yaml --inputs document_id=D0 page_num=0 -b
+```
+
+Run client with simplified yaml:
+```
+python --file pipeline.yaml --inputs document_id=D0 page_num=0
+```
+
+Run unittests:
+```
+python -m unittest discover Tests
+```
+
+Run integration test:
+```
+python integration_test.py
+```
+
+### Files:
+
+- **MLPipeline** - pipeline module
+  - **consumable** - definition of data container
+  - **pipeline** - definition of service class for pipeline
+  - **pipeline_builder** - well... pipeline builder
+  - **pipeline_task** - definition of abstract task method and task factory
+  - **pipeline_runners** - definition of specific tasks - runners
+  - **utils** - various utilities created for the project
+- **Tests** - tests module
+  - **tests** - some basic unittests
+  - **utils** - fake pipeline constructor
+- **async_pipeline_principle**.**py** - executable which demonstrate a simplified version of async pipeline run
+- **pipeline_cli**.**py** - main executable as defined in requirements
+- **requirements.txt** - requirements for project, no, I am not going to implement yaml parser, sorry guys
+- **.yaml** files - one as defined in the assignment with everything as list 'sigh', other in ordinary human form
+- **assignment.md** - assigment for the project
+
+### Flow of execution
+
+1. check tasks, if any of them have all required inputs defined (published),
+2. mark task with valid inputs as running, execute it, publish outputs, move task to finished
+3. check if global desired outputs are collected
+4. repeat until the desired outputs are collected or no more tasks could be run
+
+### Implementation options
+
+- **synchronous** - as described in flow, will be just perfect to check if the pipeline is finite (executions will be faked)
+- **asynchronous** - run all task with valid inputs, on-event of output generation check if precondition is met for other tasks (preferable if task duration ~< scheduler tick ) - *won't implement*
+- **scheduled** - run all task with available inputs asynchronously by the scheduler, easy to solve situation if task freezes or times out (preferable if task duration > scheduler tick)
```

Comment:
> easy to solve situation if task freezes or times out

True but unfortunately, it's not addressed in your code.

Reply:
True, as stated above, I am slow programer so I could not fit all requirements along with documentation of this into 1 hour. It took me 4 hours and thats what I was able to assign for this homework, certain things just did not fit the timeframe.